library/core/src/ops/deref.rs

```diff
@@ -275,6 +275,17 @@ impl<T: ?Sized> DerefMut for &mut T {
     }
 }
 
+/// UwU
```

Comment:
Not exactly sure what description we should put here. There's not exactly semantics for this trait yet.

Reply:
```suggestion
/// FIXME(deref_patterns): UwU
```

Reply:
More seriously, something like:
```
Perma-unstable marker trait. Indicates that the type has a well-behaved `Deref` (and if applicable `DerefMut`) implementation. This is relied on for soundness of deref patterns.

FIXME(deref_patterns): The precise semantics are undecided; the rough idea is that successive calls to `deref`/`deref_mut` without intermediate mutation should be idempotent, in the sense that they return the same value as far as pattern-matching is concerned.
```

Reply:
Oh, that's better

Reply:
Ah wait, I forgot something:
> Calls to `deref`/`deref_mut` must leave the pointer itself likewise unchanged.

This is to prevent a case like the hypothetical `impl Cow: DerefMut` where `deref_mut` changes the variant:
```rust
match &mut cow {
  Cow::Owned(_) => {}
  deref (ref mut _x) if false => {} // calls deref_mut, clones the value so we can mutate it
  // the value is now owned
  Cow::Borrowed(_) => {} // this doesn't match
  // the match is considered exhaustive yet none of the arms matched
}
```